packages/core/react/src/WalletProviderBase.tsx

```diff
@@ -0,0 +1,269 @@
+            adapters.map((adapter, index) => {
+                const wallet = wallets[index];
+                // If the wallet hasn't changed, return the same instance
+                return wallet && wallet.adapter === adapter && wallet.readyState === adapter.readyState
+                    ? wallet
+                    : {
+                          adapter: adapter,
+                          readyState: adapter.readyState,
+                      };
+            })
+        );
+        function handleReadyStateChange(this: Adapter, readyState: WalletReadyState) {
+            setWallets((prevWallets) => {
+                const index = prevWallets.findIndex(({ adapter }) => adapter === this);
+                if (index === -1) return prevWallets;
+
+                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                const { adapter } = prevWallets[index]!;
+                return [...prevWallets.slice(0, index), { adapter, readyState }, ...prevWallets.slice(index + 1)];
+            });
+        }
+        adapters.forEach((adapter) => adapter.on('readyStateChange', handleReadyStateChange, adapter));
+        return () => {
+            adapters.forEach((adapter) => adapter.off('readyStateChange', handleReadyStateChange, adapter));
+        };
+    }, [adapter, adapters]);
+
+    const wallet = useMemo(
+        () => wallets.find(({ adapter: candidateAdapter }) => candidateAdapter === adapter) ?? null,
+        [adapter, wallets]
+    );
+
+    // Setup and teardown event listeners when the adapter changes
+    useEffect(() => {
+        function handleWalletConnectEvent(publicKey: PublicKey) {
+            setPublicKey(publicKey);
+            isConnecting.current = false;
+            setConnecting(false);
+            setConnected(true);
+            isDisconnecting.current = false;
+            setDisconnecting(false);
+        }
+        function handleWalletDisconnectEvent() {
+            if (!isUnloadingRef.current) {
+                isConnecting.current = false;
+                setConnecting(false);
+                setConnected(false);
+                isDisconnecting.current = false;
+                setDisconnecting(false);
+                setPublicKey(null);
+            }
+        }
+        if (adapter) {
+            adapter.on('connect', handleWalletConnectEvent);
+            adapter.on('disconnect', handleWalletDisconnectEvent);
+            adapter.on('error', handleError);
+            return () => {
+                adapter.off('connect', handleWalletConnectEvent);
+                adapter.off('disconnect', handleWalletDisconnectEvent);
+                adapter.off('error', handleError);
+                isConnecting.current = false;
+                setConnecting(false);
+                setConnected(false);
+                isDisconnecting.current = false;
+                setDisconnecting(false);
+                setPublicKey(null);
+            };
+        }
```

Comment:
The original intent of combining certain state setters and only calling them when needed was to reduce the number of state updates. Does that matter, and in what circumstances will these not be correctly handled by the other places that handle this setup/teardown?

Reply:
Oh, that's not a thing in React anymore. React batches state updates by default. You can call `setFoo` 1000x in a row and you'll get 1 re-render.

Reply:
Got it. Thanks for clarifying, you have much better knowledge of React behaviors than me.